fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/AgilityContext.java

```diff
@@ -197,6 +205,16 @@ class Agile implements AgilityContext {
         private boolean committingNow = false;
         private long prevCommitCheckTime;
         private boolean closed = false;
+        private volatile Function<FDBRecordContext, CompletableFuture<Boolean>> contextPreconditions = context -> AsyncUtil.READY_TRUE;
```

Comment:
> I think when we close the FileLock, we want to remove it as a pre-condition

This answers the most important question I had. 
If we expect only few file locks per context I can just make it always return `true` when the lock is closed. `IdentityHashMap` is not thread safe and a concurrent set will be too much for this use case. What about `CopyOnWriteList`? It will be fast enough for the few entries and relatively cheap.

Reply:
Isn't that an overkill? The agility context and the lock (via directory) are bound together, so we expect only one of each. We can have a single function in agility context, set by the lock's constructor, and called by AC before every commit if non-null. I see no reason for it to be protected from multi threads as it will be set only once, and there is no other thread until the lock is achieved. 
The callback can be as simple as the lock's `ensureValid` that returns quietly if already closed. 
Something like: 
````
FDBDirectoryLock: 
   FDBDirectoryLock: 
       agilityContext.setPreCommit(this::ensureValidIfNotClosed)

AgilityContext
    setPreCommit(callback): 
       this.callback = callback 

    commitNow():  
        ....
       try {
       // already protected by semaphores     
       if (this.callback != null): 
           this.callback.accept() 
       currentContext.commit();
       currentContext.close();
      .... 
````